synthesizer/process/src/cost.rs

```diff
@@ -395,6 +416,49 @@ pub fn cost_in_microcredits<N: Network>(stack: &Stack<N>, function_name: &Identi
         })
 }
 
+/// Returns the minimum number of microcredits required to run the finalize (depcrated).
+pub fn cost_in_microcredits_deprecated<N: Network>(stack: &Stack<N>, function_name: &Identifier<N>) -> Result<u64> {
+    // Retrieve the finalize logic.
+    let Some(finalize) = stack.get_function_ref(function_name)?.finalize_logic() else {
+        // Return a finalize cost of 0, if the function does not have a finalize scope.
+        return Ok(0);
+    };
+    // Get the cost of finalizing all futures.
+    let mut future_cost = 0u64;
+    for input in finalize.inputs() {
+        if let FinalizeType::Future(future) = input.finalize_type() {
+            // Get the external stack for the future.
+            let stack = stack.get_external_stack(future.program_id())?;
+            // Accumulate the finalize cost of the future.
+            future_cost = future_cost
+                .checked_add(stack.get_finalize_cost(future.resource())?)
+                .ok_or(anyhow!("Finalize cost overflowed"))?;
+        }
+    }
+    // Aggregate the cost of all commands in the program.
+    finalize
+        .commands()
+        .iter()
+        .map(|command| cost_per_command(stack, finalize, command))
+        .try_fold(future_cost, |acc, res| {
+            res.and_then(|x| acc.checked_add(x).ok_or(anyhow!("Finalize cost overflowed")))
+        })
+        .and_then(|v2_cost| {
+        let mut additional_mapping_cost = 0u64;
+        for command in finalize.commands() {
+            match command {
+                Command::Get(_) | Command::GetOrUse(_) | Command::Contains(_) => {
+                    additional_mapping_cost = additional_mapping_cost
+                        .checked_add(MAPPING_BASE_COST_V1 - MAPPING_BASE_COST_V2)
```

Comment:
The issue is that whenever a `Stack` is `initialized`, the cost of the finalize is computed through `cost_in_microcredits` & cached: https://github.com/AleoNet/snarkVM/blob/staging/synthesizer/process/src/stack/helpers/initialize.rs#L89 . The result is that many things not currently aware of block height will have to be aware of block height, mostly unnecessarily so. For example, offline signatures & deployments (which will add a program to the process and create & initialize a new stack) will now have to be aware of the block height.

We could get rid of the caching but it seems like it was done intentionally: https://github.com/AleoNet/snarkVM/pull/2369 .

Reply:
Hmm actually on that note, won't all the **existing** stacks on running networks have cached **old fees**? So if the finalize cost is ever queried past V2 height, those would have to be updated (assuming we don't want to change the cache). Looks like this behaviour is not covered by the tests yet.

Reply:
This is a great find. It's definitely a bug but the opposite problem. Once nodes update to V2, all of the caches use new fees (as they are memory caches and don't reach rocksdb). The bug is that for imports, I was using the cached fee instead of the old higher fees. The result is that old nodes verifying transactions for nested calls with contains/get/get.or_use would have halted and the migration wouldn't have worked.

I fixed the bug and added a test here: https://github.com/AleoNet/snarkVM/pull/2564/commits/a5847dfca571f8c767425813689445936ed03690